### ISSUE TITLE
Feature Request: Pure interface for Gen

### ISSUE THREAD

**Issue by konn** [well-typed/falsify#65]
Hi,

First of all, thank you for your great library and great expositions! This library is really cool indeed.

## Feature Request: Pure interface for Gen

Currently, there is an impure interface to sample a value from `Gen`:

```hs
sample :: Gen a -> IO a
```

It would also be nice to have a *pure* analogue for this. The most straightforward one could be:

```hs
sampleFrom :: SMGen -> Gen a -> a
```

Unfortunately, this leaks the implementation detail. So perhaps we should have something like this:

```hs
data GenSource -- opaque (could be just a newtype wrapper around SMGen or SampleTree)
mkGenSource :: Word64 -> GenSource
sampleFrom :: GenSource -> Gen a -> a
```

## Background

Recently, I use `falsify` to test the validity of transformations on propositional formulae.
The test worked well, and I eventually have several valid implementations with different performance characteristics. So I decided to take some benchmarks on them.

To do that, I have to generate some random formula of a specific size. Surely, we can do this with `sample :: Gen a -> IO a`. One downside of `sample` function is the lack of reproducibility - as `sample` initialises `SMGen` from the air on every invocation, we don't have any access to the seed it used.
Once we have sampled some formulae from `Gen`, we must save them on a disk. This becomes tedious when one needs multiple formulae. If we can sample a value from specified seed, then we can just hard-code the seed instead.

**Comment by edsko** [comment]
Yes, that absolutely seems reasonable. Perhaps the library should offer a public API that can be used to define both the tasty and the ghci interfaces, and then that would then also give applications such as the one you describe what they need. I will probably not have time for this in the next few weeks though :/

**Comment by edsko** [comment]
_Technically_ speaking 3 years later is still "not in the next few weeks" 😬 But this is available as of #101, and will be released as part of `falsify-0.4` in a few weeks.
